### DISCUSSION TITLE
Milvus v1.x Clustering/HA - Single write node requirement, data consistency guarantees

### DISCUSSION THREAD

**Discussion by Neverous** [milvus-io/milvus#10215] [Q&A and General discussion]
In the Clustering docs on the site and in the blog posts I see notes that the write node is unscalable and should only be running in one copy:

[docs](https://milvus.io/docs/v1.1.1/mishards.md#Main-components) `Milvus write node: An unscalable node` 
[blog post](https://blog.milvus.io/mishards-distributed-vector-search-in-milvus-a14c239c9aad#88fd) `Only one writable node is allowed`

How much this needs to be enforced, how safe is it when potentially two writer nodes are available and accessible even for short periods of time?
From reading the docs I would assume I need to take care to make sure only one writer node is active at a time, otherwise I'm risking breaking some data (overwriting some segments potentially when two try writing at the same time? or "just" data loss from one of the instances if it gets requests?). However example configurations that are provided in the repositories only ensure that **at least** one writer node is running, they don't have any guards against running in multiple copies from what I can tell :thinking:  

How safe/reliable are they to use in production setups?

For example the [helm chart](https://github.com/milvus-io/milvus-helm/tree/1.1) uses `Deployment` for write nodes that only ensures **at least** 1 replica (I would probably expect `StatefulSet` like the databases use usually), originally I though maybe PVC setup with `accessModes: ReadWriteOnce` is intended to serve as "basic" enforcement of single write node rule (even though it only prevents multiple machines from claiming the volume, it still could potentially start multiple pods on the same machine) but it requires `ReadWriteMany` [apparently](https://github.com/milvus-io/milvus-helm/issues/100) anyway.

Likewise in bootcamp example [highly_available](https://github.com/milvus-io/bootcamp/tree/1.1/deployments/highly_available) it's using keepalived which only ensures that at least 1 node is active and in the (probably very unlikely but still) event that the cluster gets split-brained somehow with shared data storage still working, it could potentially run two instances at the same time.

Is it safe to run Milvus write node like that, does it guarantee data consistency by itself somehow?

**Comment by yhmo** [comment]
The milvus v1.x is not designed as distributed, it is more like a single instance service. We build a middleware named "[mishards](https://github.com/milvus-io/milvus/tree/1.1/shards)", but is is not a good distributed solution. 
It is also hard to build a HA service for v1.x:
1. Milvus data management is base on metadata, metadata items are not distinguished by node id, multiple writable nodes can change same metadata item. 
2. DDL/DML operations cannot be shared by multiple nodes since each node process wal log by itself. If we insert data into multiple nodes respectively, we can't ensure data consistence of the nodes.
3. The v1.x has no heartbeat. 

For 1.x, we only allow one writable node in the cluster because these disadvantages. The example [highly_available](https://github.com/milvus-io/bootcamp/tree/1.1/deployments/highly_available) is a workaround, every a writable node is crash, the keepalive will launch another milvus server. Customer also can build middleware like the Mishards according their requirement, but "one writable node" is a base rule.

Finally we completely redesign the milvus to v2.0. The 2.0 is cloud-native, we will build more features(such as HA) base on this version.
